crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1380 @@
+
+use std::{
+    cmp::min,
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use itertools::Itertools;
+use log::{debug, error, info, warn};
+use thiserror::Error as ThisError;
+
+pub type MessageHeader = u32;
+
+pub const MAX_SIMPLE_STRING_LENGTH: usize = 16; // incl. NULL terminator
+
+/// Wrapper around SCMI values of the basic types SCMI defines.
+///
+/// Everything communicating to/from SCMI must be composed of them.
+// SCMI specification talks about Le32 parameter and return values.
+// VirtIO SCMI specification talks about u8 SCMI values.
+// Let's stick with SCMI specification for implementation simplicity.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MessageValue {
+    Signed(i32),
+    Unsigned(u32),
+    String(String, usize), // string, expected characters
+}
+
+impl MessageValue {
+    pub(crate) fn get_unsigned(&self) -> u32 {
+        match self {
+            Self::Unsigned(value) => *value,
+            _ => panic!("Wrong parameter"),
+        }
+    }
+}
+
+pub type MessageValues = Vec<MessageValue>;
+
+/// Enumeration of SCMI message types, mapped to the corresponding SCMI codes.
+///
+/// The only one we currently support is [MessageType::Command].
+#[derive(Debug, PartialEq)]
+enum MessageType {
+    // 4-bit unsigned integer
+    Command,     // 0
+    Unsupported, // anything else
+}
+pub type MessageId = u8;
+pub type ProtocolId = u8;
+type NParameters = u8;
+
+/// Mapping of return values to SCMI return status codes.
+#[derive(Clone, Copy)]
+#[allow(dead_code)]
+enum ReturnStatus {
+    // 32-bit signed integer
+    Success = 0,
+    NotSupported = -1,
+    InvalidParameters = -2,
+    Denied = -3,
+    NotFound = -4,
+    OutOfRange = -5,
+    Busy = -6,
+    CommsError = -7,
+    GenericError = -8,
+    HardwareError = -9,
+    ProtocolError = -10,
+    // -11..-127: reserved
+    // <-127: vendor specific
+}
+
+impl ReturnStatus {
+    const fn as_value(&self) -> MessageValue {
+        MessageValue::Signed(*self as i32)
+    }
+}
+
+/// Representation of [MessageValue] sequence used to construct [ScmiResponse].
+///
+/// The sequence includes the response code (see the helper constructors for
+/// adding them) but it doesn't include the SCMI message header.  The header is
+/// added in [ScmiResponse].
+struct Response {
+    values: MessageValues,
+}
+
+impl From<ReturnStatus> for Response {
+    fn from(value: ReturnStatus) -> Self {
+        Self {
+            values: vec![value.as_value()],
+        }
+    }
+}
+
+impl From<MessageValue> for Response {
+    fn from(value: MessageValue) -> Self {
+        Self {
+            values: vec![ReturnStatus::Success.as_value(), value],
+        }
+    }
+}
+
+impl From<&mut MessageValues> for Response {
+    fn from(value: &mut MessageValues) -> Self {
+        let mut response_values = vec![ReturnStatus::Success.as_value()];
+        response_values.append(value);
+        Self {
+            values: response_values,
+        }
+    }
+}
```

Comment:
```suggestion
impl From<MessageValues> for Response {
    fn from(value: MessageValues) -> Self {
        let mut response_values = vec![ReturnStatus::Success.as_value()];
        response_values.append(value);
        Self {
            values: response_values,
        }
    }
}
```
The &mut is not needed

Reply:
The compiler thinks otherwise because `append`'s argument is &mut.

Reply:
Do you want to explicitly drain a mutable `MessageValues` reference (which is a typedef for a `Vec<MessageValue>`) here? That' what `append` does. Otherwise, you can just do: 


```suggestion
impl From<&[MessageValue]> for Response {
    fn from(value: &[MessageValue]) -> Self {
        let mut response_values = vec![ReturnStatus::Success.as_value()];
        response_values.extend_from_slice(value);
        Self {
            values: response_values,
        }
    }
}
```

Reply:
I see, thanks, done.